File: src/subsystems/coral_arm.py
import commands2
import rev
import wpilib
import wpimath.controller
from wpilib.sysid import SysIdRoutineLog
from commands2.sysid import SysIdRoutine
from wpilib import RobotController
from wpilib.simulation import SingleJointedArmSim, RoboRioSim
from wpimath.geometry import Rotation2d
from wpimath.system.plant import DCMotor, LinearSystemId
from wpiutil import SendableBuilder

from constants import CoralArmConstants
from sim_helper import SimHelper


class CoralArm(commands2.Subsystem):
    def __init__(self):
        super().__init__()
        self.setName("Coral Arm")

        # Setup motors
        self.motor = rev.SparkFlex(CoralArmConstants.MOTOR_ID, rev.SparkBase.MotorType.kBrushless)
        self.absolute_encoder = self.motor.getAbsoluteEncoder()  # REV Through Bore Encoder
        self.controller = self.motor.getClosedLoopController()

        self.feedforward = wpimath.controller.ArmFeedforward(*CoralArmConstants.FEEDFORWARD_CONSTANTS)

        # Use an encoder offset on the real robot, but don't offset the encoder in simulation
        self.encoder_offset = CoralArmConstants.ENCODER_OFFSET if wpilib.RobotBase.isReal() else 0

        self.config()

        # Setup mechanism and gearbox for simulation
        self.absolute_encoder_sim = rev.SparkAbsoluteEncoderSim(self.motor)
        gearbox = DCMotor.neoVortex(1)
        self.motor_sim = rev.SparkFlexSim(self.motor, gearbox)
        moment = SingleJointedArmSim.estimateMOI(CoralArmConstants.ARM_LENGTH, CoralArmConstants.ARM_MASS)
        plant = LinearSystemId.singleJointedArmSystem(gearbox, moment, CoralArmConstants.GEAR_RATIO)
        self.arm_sim = SingleJointedArmSim(
            plant,
            gearbox,
            CoralArmConstants.ARM_MASS,
            CoralArmConstants.ARM_MASS,
            CoralArmConstants.MINIMUM_ANGLE.radians(),
            CoralArmConstants.MAXIMUM_ANGLE.radians(),
            True,
            0,
        )

        # Visual display of the arm
        mech = wpilib.Mechanism2d(5, 5)
        root = mech.getRoot("armPivot", 1, 2.5)
        self.arm = root.appendLigament("arm", 3, self.arm_sim.getAngleDegrees())

        # Create a new SysId routine for characterizing the arm
        self.sysId_routine = SysIdRoutine(
            SysIdRoutine.Config(0.25, 3),
            SysIdRoutine.Mechanism(
                self.set_voltage,
                self.sysId_log,
                self,
            ),
        )

        # Reset NEO encoder to absolute encoder
        self.motor.getEncoder().setPosition(
            self.absolute_encoder.getPosition()
        )

        wpilib.SmartDashboard.putData("Arm Mechanism", mech)

    def simulationPeriodic(self) -> None:
        self.arm_sim.setInputVoltage(self.motor_sim.getAppliedOutput() * RoboRioSim.getVInVoltage())

        self.arm_sim.update(0.02)

        self.motor_sim.iterate(
            self.arm_sim.getVelocityDps(),
            RoboRioSim.getVInVoltage(),
            0.02
        )

        self.absolute_encoder_sim.iterate(self.arm_sim.getVelocityDps(), 0.02)

        SimHelper.add_simulated_current_load(RobotController.getTime(), self.arm_sim.getCurrentDraw())

        self.arm.setAngle(self.arm_sim.getAngleDegrees())

    def config(self):
        motor_config = rev.SparkBaseConfig()

        motor_config \
            .setIdleMode(rev.SparkBaseConfig.IdleMode.kBrake) \

        motor_config.absoluteEncoder \
            .positionConversionFactor(360) \
            .velocityConversionFactor(360 / 60)

        motor_config.encoder \
            .positionConversionFactor(360 / CoralArmConstants.GEAR_RATIO) \
            .velocityConversionFactor(360 / CoralArmConstants.GEAR_RATIO / 60)

        motor_config.closedLoop \
            .setFeedbackSensor(rev.ClosedLoopConfig.FeedbackSensor.kAbsoluteEncoder) \
            .pid(CoralArmConstants.kP, 0, 0) \
            .outputRange(-1, 1) \

        motor_config.softLimit \
            .forwardSoftLimit(CoralArmConstants.MAXIMUM_ANGLE.degrees() + self.encoder_offset) \
            .reverseSoftLimit(CoralArmConstants.MINIMUM_ANGLE.degrees() + self.encoder_offset) \
            .forwardSoftLimitEnabled(True) \
            .reverseSoftLimitEnabled(True) \

        self.motor.configure(
            motor_config,
            rev.SparkBase.ResetMode.kResetSafeParameters,
            rev.SparkBase.PersistMode.kPersistParameters,
        )

    def set_angle(self, angle: Rotation2d):
        """
        Command the arm to move to a specific angle. 0 degrees is considered horizontal, facing toward
        the front of the robot.

        :param angle: Counter-clockwise positive angle where 0 degrees is horizontal, facing the front of the robot.
        """
        ff = self.feedforward.calculate(self.angle().degrees(), self.absolute_encoder.getVelocity())
        self.controller.setReference(
            angle.degrees() + self.encoder_offset,
            rev.SparkBase.ControlType.kPosition,
            arbFeedforward=ff,
            arbFFUnits=rev.SparkClosedLoopController.ArbFFUnits.kVoltage,
        )

    def set_duty_cycle(self, output: float):
        self.motor.set(output)

    def at_rotation(self, goal_rotation: Rotation2d) -> bool:
        return (goal_rotation - CoralArmConstants.ARM_TOLERANCE).radians() < self.angle().radians() < (goal_rotation + CoralArmConstants.ARM_TOLERANCE).radians()

    def set_voltage(self, volts: float):
        self.controller.setReference(volts, rev.SparkBase.ControlType.kVoltage)

    def angle(self) -> Rotation2d:
        degrees = self.absolute_encoder.getPosition() - self.encoder_offset
        return Rotation2d.fromDegrees(degrees)

    def sysId_log(self, log: SysIdRoutineLog):
        log.motor("arm-pivot") \
            .voltage(self.motor.getAppliedOutput() * self.motor.getBusVoltage()) \
            .angularPosition(self.angle().degrees()) \
            .angularVelocity(self.absolute_encoder.getVelocity())

    def initSendable(self, builder: SendableBuilder) -> None:
        builder.addStringProperty("Command", self.current_command_name, lambda _: None)
        builder.addDoubleProperty("Angle", lambda: self.angle().degrees(), lambda degrees: self.set_angle(Rotation2d.fromDegrees(degrees)))

    def current_command_name(self) -> str:
        try:
            return self.getCurrentCommand().getName()
        except AttributeError:
            return ""

    def SetAngleCommand(self, angle: Rotation2d):
        return commands2.RunCommand(lambda: self.set_angle(angle), self) \
            .until(lambda: self.at_rotation(angle))

    def SysIdQuasistatic(self, direction: SysIdRoutine.Direction):
        return self.sysId_routine.quasistatic(direction)

    def SysIdDynamic(self, direction: SysIdRoutine.Direction):
        return self.sysId_routine.dynamic(direction)

File: src_java/frc/robot/subsystems/CoralArm.java
package frc.robot.subsystems;

import com.revrobotics.AbsoluteEncoder;
import com.revrobotics.spark.ClosedLoopSlot;
import com.revrobotics.spark.SparkBase.ControlType;
import com.revrobotics.spark.SparkBase.PersistMode;
import com.revrobotics.spark.SparkBase.ResetMode;
import com.revrobotics.spark.SparkClosedLoopController;
import com.revrobotics.spark.SparkFlex;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.config.ClosedLoopConfig.FeedbackSensor;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.config.SparkFlexConfig;

import edu.wpi.first.math.controller.ArmFeedforward;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.smartdashboard.Mechanism2d;
import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine;
import frc.robot.Constants.CoralArmConstants;

import static edu.wpi.first.units.Units.*;

/**
 * Coral arm subsystem for rotating the scoring mechanism.
 * Uses a NEO Vortex motor with an absolute encoder.
 */
public class CoralArm extends SubsystemBase {
    // Motor and encoder
    private final SparkFlex motor;
    private final AbsoluteEncoder absoluteEncoder;
    private final SparkClosedLoopController controller;

    // Feedforward
    private final ArmFeedforward feedforward;

    // Encoder offset for the real robot
    private final double encoderOffset;

    // Mechanism visualization
    private final Mechanism2d mechanism;
    private final MechanismLigament2d armLigament;

    // SysId routine
    private final SysIdRoutine sysIdRoutine;

    // Goal tracking
    private Rotation2d goalRotation = null;

    /**
     * Creates a new CoralArm subsystem.
     */
    public CoralArm() {
        // Initialize motor
        motor = new SparkFlex(CoralArmConstants.MOTOR_ID, MotorType.kBrushless);
        absoluteEncoder = motor.getAbsoluteEncoder();
        controller = motor.getClosedLoopController();

        // Initialize feedforward
        feedforward = new ArmFeedforward(
            CoralArmConstants.FEEDFORWARD_CONSTANTS[0],  // kS
            CoralArmConstants.FEEDFORWARD_CONSTANTS[1],  // kG
            CoralArmConstants.FEEDFORWARD_CONSTANTS[2],  // kV
            CoralArmConstants.FEEDFORWARD_CONSTANTS[3]   // kA
        );

        // Use encoder offset on real robot, but don't offset in simulation
        encoderOffset = RobotBase.isReal() ? CoralArmConstants.ENCODER_OFFSET : 0;

        // Configure motor
        configureMotor();

        // Reset NEO encoder to absolute encoder position
        motor.getEncoder().setPosition(absoluteEncoder.getPosition());

        // Setup mechanism visualization
        mechanism = new Mechanism2d(5, 5);
        MechanismRoot2d root = mechanism.getRoot("armPivot", 1, 2.5);
        armLigament = root.append(new MechanismLigament2d("arm", 3, getAngle().getDegrees()));
        SmartDashboard.putData("Arm Mechanism", mechanism);

        // Setup SysId routine
        sysIdRoutine = new SysIdRoutine(
            new SysIdRoutine.Config(
                Volts.of(0.25).per(Seconds.of(1)),
                Volts.of(3),
                null,
                null
            ),
            new SysIdRoutine.Mechanism(
                voltage -> setVoltage(voltage.in(Volts)),
                log -> {
                    log.motor("arm-pivot")
                        .voltage(Volts.of(motor.getAppliedOutput() * motor.getBusVoltage()))
                        .angularPosition(Degrees.of(getAngle().getDegrees()))
                        .angularVelocity(DegreesPerSecond.of(absoluteEncoder.getVelocity()));
                },
                this
            )
        );
    }

    /**
     * Configure the motor with appropriate settings.
     */
    private void configureMotor() {
        SparkFlexConfig config = new SparkFlexConfig();

        config.idleMode(IdleMode.kBrake);

        // Configure absolute encoder
        config.absoluteEncoder
            .positionConversionFactor(360)
            .velocityConversionFactor(360.0 / 60.0);

        // Configure relative encoder
        config.encoder
            .positionConversionFactor(360.0 / CoralArmConstants.GEAR_RATIO)
            .velocityConversionFactor(360.0 / CoralArmConstants.GEAR_RATIO / 60.0);

        // Configure closed-loop control
        config.closedLoop
            .feedbackSensor(FeedbackSensor.kAbsoluteEncoder)
            .p(CoralArmConstants.kP)
            .i(0)
            .d(0)
            .outputRange(-1, 1);

        // Configure soft limits
        config.softLimit
            .forwardSoftLimit(CoralArmConstants.MAXIMUM_ANGLE.getDegrees() + encoderOffset)
            .reverseSoftLimit(CoralArmConstants.MINIMUM_ANGLE.getDegrees() + encoderOffset)
            .forwardSoftLimitEnabled(true)
            .reverseSoftLimitEnabled(true);

        motor.configure(config, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    }

    @Override
    public void periodic() {
        // Update mechanism visualization
        armLigament.setAngle(getAngle().getDegrees());

        // Log data
        SmartDashboard.putNumber("CoralArm/Angle", getAngle().getDegrees());
        SmartDashboard.putNumber("CoralArm/Velocity", absoluteEncoder.getVelocity());
        SmartDashboard.putBoolean("CoralArm/AtGoal", atGoalRotation());
        SmartDashboard.putString("CoralArm/Command", getCurrentCommandName());

        if (goalRotation != null) {
            SmartDashboard.putNumber("CoralArm/GoalAngle", goalRotation.getDegrees());
        }
    }

    /**
     * Set the arm to a specific angle.
     * 0 degrees is horizontal, facing toward the front of the robot.
     *
     * @param angle Counter-clockwise positive angle
     */
    public void setAngle(Rotation2d angle) {
        goalRotation = angle;
        controller.setReference(
            angle.getDegrees() + encoderOffset,
            ControlType.kPosition
        );
    }

    /**
     * Set the motor duty cycle directly.
     *
     * @param output The duty cycle (-1 to 1)
     */
    public void setDutyCycle(double output) {
        motor.set(output);
    }

    /**
     * Set the motor voltage directly.
     *
     * @param volts The voltage to apply
     */
    public void setVoltage(double volts) {
        controller.setReference(volts, ControlType.kVoltage);
    }

    /**
     * Check if the arm is at a specific rotation.
     *
     * @param rotation The rotation to check
     * @return True if at the target rotation within tolerance
     */
    public boolean atRotation(Rotation2d rotation) {
        double diff = Math.abs(getAngle().getRadians() - rotation.getRadians());
        return diff < CoralArmConstants.ARM_TOLERANCE.getRadians();
    }

    /**
     * Check if the arm is at the goal rotation.
     *
     * @return True if at the goal rotation
     */
    public boolean atGoalRotation() {
        return goalRotation != null && atRotation(goalRotation);
    }

    /**
     * Get the current angle of the arm.
     *
     * @return The current angle as a Rotation2d
     */
    public Rotation2d getAngle() {
        double degrees = absoluteEncoder.getPosition() - encoderOffset;
        return Rotation2d.fromDegrees(degrees);
    }

    /**
     * Get the name of the current command.
     *
     * @return The current command name or empty string
     */
    public String getCurrentCommandName() {
        Command current = getCurrentCommand();
        return current != null ? current.getName() : "";
    }

    // ==================== Commands ====================

    /**
     * Command to set the arm to a specific angle.
     *
     * @param angle The target angle
     * @return A command that moves the arm to the angle
     */
    public Command setAngleCommand(Rotation2d angle) {
        return Commands.run(() -> setAngle(angle), this)
            .until(this::atGoalRotation)
            .withName("SetAngle(" + angle.getDegrees() + ")");
    }

    /**
     * SysId quasistatic command.
     *
     * @param direction The direction to run
     * @return The SysId command
     */
    public Command sysIdQuasistatic(SysIdRoutine.Direction direction) {
        return sysIdRoutine.quasistatic(direction);
    }

    /**
     * SysId dynamic command.
     *
     * @param direction The direction to run
     * @return The SysId command
     */
    public Command sysIdDynamic(SysIdRoutine.Direction direction) {
        return sysIdRoutine.dynamic(direction);
    }
}
